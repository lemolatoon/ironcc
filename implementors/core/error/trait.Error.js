(function() {var implementors = {
"ironcc":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.67.0/core/error/trait.Error.html\" title=\"trait core::error::Error\">Error</a> for <a class=\"struct\" href=\"ironcc/error/struct.CompileError.html\" title=\"struct ironcc::error::CompileError\">CompileError</a>"]]
};if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()