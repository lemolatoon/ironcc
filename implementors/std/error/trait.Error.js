(function() {var implementors = {};
implementors["ironcc"] = [{"text":"impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.64.0/std/error/trait.Error.html\" title=\"trait std::error::Error\">Error</a> for <a class=\"struct\" href=\"ironcc/error/struct.CompileError.html\" title=\"struct ironcc::error::CompileError\">CompileError</a>","synthetic":false,"types":["ironcc::error::CompileError"]}];
if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()