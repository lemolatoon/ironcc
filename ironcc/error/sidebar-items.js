window.SIDEBAR_ITEMS = {"enum":[["AnalyzeErrorKind",""],["CompileErrorKind",""],["GenerateErrorKind",""],["NotAllowedStmtKind",""],["ParseErrorKind",""],["TokenizeErrorKind",""],["Tokens",""],["TypeErrorKind",""],["TypeExpectedFailedKind",""],["UnexpectedTypeSizeStatus",""],["VariableKind",""]],"struct":[["CompileError",""]]};