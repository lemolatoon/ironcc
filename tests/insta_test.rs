use insta::assert_debug_snapshot;
pub mod test_utils;
use ironcc::error::CompileError;
use test_utils::CachedProcessor;

macro_rules! all {
    ($self: ident) => {
        assert_debug_snapshot!($self.tokens().as_ref().unwrap());
        assert_debug_snapshot!($self.program().as_ref().unwrap());
        assert_debug_snapshot!($self.conv_program().as_ref().unwrap());
    };
}

#[test]
fn insta_tests() -> Result<(), CompileError> {
    let src = "\nint main() {\nint i;\ni = 5;\nint* p; p = &i;\nint *p2; p2 = p + i;\n}";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn initializer() -> Result<(), CompileError> {
    let src = "int main() {int a = 5; int *p = &a; return 0;}";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn size_of() -> Result<(), CompileError> {
    let src = "\n\
    int main() {\n \
        int a = 5;\n \
        int b = sizeof (a);\n \
        int c = sizeof (int);\n \
        int d = sizeof (int*);\n \
    }\n\
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn array_syntax_sugar() -> Result<(), CompileError> {
    let src = "\n\
    int main() {\n \
        int array[5];
        return array[0];
    }\n\
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn scopes() {
    let src = "\n\
    int main() {\n \
        int a;
        {int *k;}
        {int *i; {int k;}}
        int c[2];
    }\n\
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn implicit_cast() {
    let src = "
    int main() {
        int a = 2;
        char b = a;
        b = 4;
    }
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn add_char() {
    let src = "
    int main() {
        int a = 2;
        char b = a;
        a + b;
        b + a;
    }
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn string_literal() {
    let src = r#"
    int printf(char *msg);

    int main() {
        char *msg = "Hello World";
        printf(msg);
    }
    "#;
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn string_literal2() {
    let src = r#"
    int printf(char *msg);

    int main() {
        char *msg;
        msg = "Hello World";
        printf(msg);
    }
    "#;
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}
