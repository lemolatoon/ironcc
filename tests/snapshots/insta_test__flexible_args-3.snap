---
source: tests/insta_test.rs
expression: tester.conv_program().as_ref().unwrap()
---
ConvProgram {
    components: [
        Global(
            GVar {
                name: ".LC0",
                ty: Array(
                    Base(
                        Char,
                    ),
                    16,
                ),
                init: Some(
                    Array(
                        [
                            ConstExpr {
                                kind: Int(
                                    72,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    101,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    108,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    108,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    111,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    32,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    87,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    111,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    114,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    108,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    100,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    58,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    37,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    100,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    10,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                            ConstExpr {
                                kind: Int(
                                    0,
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 15,
                                    n_line: 5,
                                },
                            },
                        ],
                    ),
                ),
            },
        ),
        Func(
            ConvFuncDef {
                ty: Func {
                    ret_ty: Base(
                        Int,
                    ),
                    args: [],
                    is_flexible: true,
                },
                name: "main",
                args: [],
                body: Block(
                    [
                        Expr(
                            ConvExpr {
                                kind: Assign(
                                    ConvExpr {
                                        kind: LVar(
                                            LVar {
                                                offset: 4,
                                                ty: Base(
                                                    Int,
                                                ),
                                            },
                                        ),
                                        ty: Base(
                                            Int,
                                        ),
                                        pos: Position {
                                            n_char: 8,
                                            n_line: 4,
                                        },
                                    },
                                    ConvExpr {
                                        kind: Num(
                                            2,
                                        ),
                                        ty: Base(
                                            Int,
                                        ),
                                        pos: Position {
                                            n_char: 16,
                                            n_line: 4,
                                        },
                                    },
                                ),
                                ty: Base(
                                    Int,
                                ),
                                pos: Position {
                                    n_char: 8,
                                    n_line: 4,
                                },
                            },
                        ),
                        Expr(
                            ConvExpr {
                                kind: Func(
                                    "printf",
                                    [
                                        ConvExpr {
                                            kind: Addr(
                                                ConvExpr {
                                                    kind: GVar(
                                                        GVar {
                                                            name: ".LC0",
                                                            ty: Array(
                                                                Base(
                                                                    Char,
                                                                ),
                                                                16,
                                                            ),
                                                            init: Some(
                                                                Array(
                                                                    [
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                72,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                101,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                108,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                108,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                111,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                32,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                87,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                111,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                114,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                108,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                100,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                58,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                37,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                100,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                10,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                        ConstExpr {
                                                                            kind: Int(
                                                                                0,
                                                                            ),
                                                                            ty: Base(
                                                                                Int,
                                                                            ),
                                                                            pos: Position {
                                                                                n_char: 15,
                                                                                n_line: 5,
                                                                            },
                                                                        },
                                                                    ],
                                                                ),
                                                            ),
                                                        },
                                                    ),
                                                    ty: Base(
                                                        Char,
                                                    ),
                                                    pos: Position {
                                                        n_char: 15,
                                                        n_line: 5,
                                                    },
                                                },
                                            ),
                                            ty: Ptr(
                                                Base(
                                                    Char,
                                                ),
                                            ),
                                            pos: Position {
                                                n_char: 15,
                                                n_line: 5,
                                            },
                                        },
                                        ConvExpr {
                                            kind: LVar(
                                                LVar {
                                                    offset: 4,
                                                    ty: Base(
                                                        Int,
                                                    ),
                                                },
                                            ),
                                            ty: Base(
                                                Int,
                                            ),
                                            pos: Position {
                                                n_char: 35,
                                                n_line: 5,
                                            },
                                        },
                                    ],
                                    0,
                                ),
                                ty: Void,
                                pos: Position {
                                    n_char: 8,
                                    n_line: 5,
                                },
                            },
                        ),
                    ],
                ),
                stack_size: 4,
            },
        ),
    ],
}
